### PR TITLE
[Flang][OpenMP] Derived type explicit allocatable member mapping

### PR DESCRIPTION
**Pull request by agozillon**
This PR is one of 3 in a PR stack, this is the primary change set which seeks
to extend the current derived type explicit member mapping support to
handle descriptor member mapping at arbitrary levels of nesting. The PR
stack seems to do this reasonably (from testing so far) but as you can
create quite complex mappings with derived types (in particular when adding
allocatable derived types or arrays of allocatable derived types) I imagine
there will be hiccups, which I am more than happy to address. There will
also be further extensions to this work to handle the implicit auto-magical
mapping of descriptor members in derived types and a few other changes
planned for the future (with some ideas on optimizing things).

The changes in this PR primarily occur in the OpenMP lowering and
the OMPMapInfoFinalization pass.

In the OpenMP lowering several utility functions were added or extended
to support the generation of appropriate intermediate member mappings
which are currently required when the parent (or multiple parents) of a
mapped member are descriptor types. We need to map the entirety of
these types or do a "deep copy" for lack of a better term, where we map
both the base address and the descriptor as without the copying of both
of these we lack the information in the case of the descriptor to access the
member or attach the pointers data to the pointer and in the latter case we
require the base address to map the chunk of data. Currently we do not
segment descriptor based derived types as we do with regular
non-descriptor derived types, we effectively map their entirety in all
cases at the moment, I hope to address this at some point in the future as
it adds a fair bit of a performance penalty to having nestings of allocatable
derived types as an example. The process of mapping all intermediate
descriptor members in a members path only occurs if a member has
an allocatable or object parent in its symbol path or the member itself
is a member or allocatable. This occurs in the
createParentSymAndGenIntermediateMaps function, which will also
generate the appropriate address for the allocatable member
within the derived type to use as a the varPtr field of the map (for
intermediate allocatable maps and final allocatable mappings). In
this case it's necessary as we can't utilise the usual Fortran::lower
functionality such as gatherDataOperandAddrAndBounds without
causing issues later in the lowering due to extra allocas being spawned
which seem to affect the pointer attachment (at least this is my
current assumption, it results in memory access errors on the device
due to incorrect map information generation). This is similar
to why we do not use the MLIR value generated for this and utilise
the original symbol provided when mapping descriptor types external
to derived types. Hopefully this can be rectified in the future so this
function can be simplified and more closely aligned to the other type
mappings. We also make use of fir::CoordinateOp as opposed to the
HLFIR version as the HLFIR version doesn't support the appropriate
lowering to FIR necessary at the moment, we also cannot use a
single CoordinateOp (similarly to a single GEP) as when we index
through a descriptor operation (BoxType) we encounter issues later
in the lowering, however in either case we need access to intermediate
descriptors so individual CoordinateOp's aid this (although, being
able to compress them into a smaller amount of CoordinateOp's may
simplify the IR and perhaps result in a better end product, something
to consider for the future).

The other large change area was in the OMPMapInfoFinalization pass,
where the pass had to be extended to support the expansion of box
types (or multiple nestings of box types) within derived types, or box
type derived types. This requires expanding each BoxType mapping
from one into two maps and then modifying all of the existing
member indices of the overarching parent mapping to account for
the addition of these new members alongside adjusting the existing
member indices to support the addition of these new maps which
extend the original member indices (as a base address of a box type
is currently considered a member of the box type at a position of
0 as when lowered to LLVM-IR it's a pointer contained at this position
in the descriptor type, however, this means extending mapped children
of this expanded descriptor type to additionally incorporate the new
member index in the correct location in its own index list). I believe
there is a reasonable amount of comments that should aid in
understanding this better, alongside the test alterations for the pass.

A subset of the changes were also aimed at making some of the utilities
for packing and unpacking the DenseIntElementsAttr
containing the member indices shareable across the lowering and
OMPMapInfoFinalization, this required moving some functions to the
Lower/Support/Utils.h header, and transforming the lowering structure
containing the member index data into something more similar to the
version used in OMPMapInfoFinalization. There we also some other
attempts at tidying things up in relation to the member index data
generation in the lowering, some of which required creating a logical
operator for the OpenMP ID class so it can be utilised as a map key
(it simply utilises the symbol address for the moment as ordering
isn't particularly important).

Otherwise I have added a set of new tests encompassing some of
the mappings currently supported by this PR (unfortunately as
you can have arbitrary nestings of all shapes and types it's not
very feasible to cover them all).

### CODE REVIEW

**Comment by agozillon** [comment]
This portion of the PR stack underwent the most changes, as it's the most "complex" part of the PR stack, it's primarily underwent some additional complexity in the map lowering for member mapping which now attempts to incorporate bounds where possible amongst a some other things, it also incorporates the last iteration of reviewer comments on the previous PR stack alongside additional tests checking the new changes work as expected (and continue to).

File: flang/test/Lower/OpenMP/derived-type-allocatable-map.f90 [hunk 1]
```diff
@@ -0,0 +1,161 @@
+!RUN: %flang_fc1 -emit-hlfir -fopenmp %s -o - | FileCheck %s
```
Comment (ergawy) [inline flang/test/Lower/OpenMP/derived-type-allocatable-map.f90:1]: In the whole file, can you capture the names of longs types in FileCheck variables and use the captured type names to improve readability and reduce line lengths a bit?

File: flang/test/Lower/OpenMP/derived-type-allocatable-map.f90 [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
...perand_and_block_additionTmiddle_layer{i:f32,array_i:!fir.array<10xi32>,array_k:!fir.box<!fir.heap<!fir.array<?xi32>>>,k:i32}>}>>>) map_clauses(to) capture(ByRef) members(%75, %69, %68, %73 : [0], [0,6,2], [0,6,2,0], [0,6,3] : !fir.llvm_ptr<!fir.ref<!fir.type<_QFtest_alloca_nested_derived_type_map_operand_and_block_additionTtop_layer{i:f32,scalar:!fir.box<!fir.heap<i32>>,array_i:!fir.array<10xi32>,j:f32,array_j:!fir.box<!fir.heap<!fir.array<?xi32>>>,k:i32,nest:!fir.type<_QFtest_alloca_nested_derived_type_map_operand_and_block_additionTmiddle_layer{i:f32,array_i:!fir.array<10xi32>,array_k:!fir.box<!fir.heap<!fir.array<?xi32>>>,k:i32}>}>>>, !fir.ref<!fir.box<!fir.heap<!fir.array<?xi32>>>>, !fir.llvm_ptr<!fir.ref<!fir.array<?xi32>>>, !fir.ref<i32>) -> !fir.ref<!fir.box<!fir.heap<!fir.type<_QFtest_alloca_nested_derived_type_map_operand_and_block_additionTtop_layer{i:f32,scalar:!fir.box<!fir.heap<i32>>,array_i:!fir.array<10xi32>,j:f32,array_j:!fir.box<!fir.heap<!fir.array<?xi32>>>,k:i32,nest:!fir.type<_QFtest_alloca_nested_derived_type_map_operand_and_block_additionTmiddle_layer{i:f32,array_i:!fir.array<10xi32>,array_k:!fir.box<!fir.heap<!fir.array<?xi32>>>,k:i32}>}>>>> {name = "one_l"}
```
Comment (ergawy) [inline flang/test/Lower/OpenMP/derived-type-allocatable-map.f90:91]: I think you forgot to use the FileCheck vars for `members` here.

File: flang/test/Lower/OpenMP/derived-type-allocatable-map.f90 [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
...2>>>,k:i32,nest:!fir.type<_QFtest_nested_derived_type_alloca_map_operand_and_block_additionTmiddle_layer{i:f32,array_i:!fir.array<10xi32>,array_k:!fir.box<!fir.heap<!fir.array<?xi32>>>,k:i32}>}>>, !fir.type<_QFtest_nested_derived_type_alloca_map_operand_and_block_additionTtop_layer{i:f32,scalar:!fir.box<!fir.heap<i32>>,array_i:!fir.array<10xi32>,j:f32,array_j:!fir.box<!fir.heap<!fir.array<?xi32>>>,k:i32,nest:!fir.type<_QFtest_nested_derived_type_alloca_map_operand_and_block_additionTmiddle_layer{i:f32,array_i:!fir.array<10xi32>,array_k:!fir.box<!fir.heap<!fir.array<?xi32>>>,k:i32}>}>) map_clauses(tofrom) capture(ByRef) members(%62, %61 : [6,2], [6,2,0] : !fir.ref<!fir.box<!fir.heap<!fir.array<?xi32>>>>, !fir.llvm_ptr<!fir.ref<!fir.array<?xi32>>>) -> !fir.ref<!fir.type<_QFtest_nested_derived_type_alloca_map_operand_and_block_additionTtop_layer{i:f32,scalar:!fir.box<!fir.heap<i32>>,array_i:!fir.array<10xi32>,j:f32,array_j:!fir.box<!fir.heap<!fir.array<?xi32>>>,k:i32,nest:!fir.type<_QFtest_nested_derived_type_alloca_map_operand_and_block_additionTmiddle_layer{i:f32,array_i:!fir.array<10xi32>,array_k:!fir.box<!fir.heap<!fir.array<?xi32>>>,k:i32}>}>> {name = "one_l", partial_map = true}
```
Comment (ergawy) [inline flang/test/Lower/OpenMP/derived-type-allocatable-map.f90:133]: Same here.

**Review by ergawy — COMMENTED** [review]
First round, just reading through some tests to make sure I understand the effect of the changes. Will continue later. Thanks Andrew!

File: flang/lib/Lower/OpenMP/Utils.cpp [hunk 4]
```diff
@@ -145,11 +146,174 @@ createMapInfoOp(fir::FirOpBuilder &builder, mlir::Location loc,
       builder.getIntegerAttr(builder.getIntegerType(64, false), mapType),
       builder.getAttr<mlir::omp::VariableCaptureKindAttr>(mapCaptureType),
       builder.getStringAttr(name), builder.getBoolAttr(partialMap));
-
   return op;
 }
 
-static int
+omp::ObjectList gatherObjects(omp::Object obj,
```
Comment (ergawy) [inline flang/lib/Lower/OpenMP/Utils.cpp:152]:
```suggestion
omp::ObjectList gatherParentObjectsOf(omp::Object derivedTypeMember,
```

File: flang/lib/Lower/OpenMP/Utils.cpp [hunk 5]
```diff
@@ -145,11 +146,174 @@ createMapInfoOp(fir::FirOpBuilder &builder, mlir::Location loc,
...
+  omp::ObjectList objList;
+  std::optional<omp::Object> baseObj = obj;
+  while (baseObj.has_value()) {
+    objList.push_back(baseObj.value());
+    baseObj = getBaseObject(baseObj.value(), semaCtx);
+  }
+  return omp::ObjectList{llvm::reverse(objList)};
+}
+
+bool isDuplicateMemberMapInfo(OmpMapParentAndMemberData &parentMembers,
```
Comment (ergawy) [inline flang/lib/Lower/OpenMP/Utils.cpp:163]: Can this be a method in `OmpMapParentAndMemberData`?

File: flang/lib/Lower/OpenMP/Utils.cpp [hunk 6]
```diff
@@ -145,11 +146,174 @@ createMapInfoOp(fir::FirOpBuilder &builder, mlir::Location loc,
...
+  return false;
+}
+
+static void generateArrayIndices(lower::AbstractConverter &converter,
+                                 fir::FirOpBuilder &firOpBuilder,
+                                 lower::StatementContext &stmtCtx,
+                                 mlir::Location clauseLocation,
+                                 llvm::SmallVectorImpl<mlir::Value> &indices,
+                                 omp::Object object) {
+  if (auto maybeRef = evaluate::ExtractDataRef(*object.ref())) {
```
Comment (ergawy) [inline flang/lib/Lower/OpenMP/Utils.cpp:178]: Can you use early returns for the first 2 `if`s and `contine` without `else` for the 3rd?

File: flang/lib/Lower/OpenMP/Utils.cpp [hunk 7]
```diff
@@ -145,11 +146,174 @@ createMapInfoOp(fir::FirOpBuilder &builder, mlir::Location loc,
...
+bool isDuplicateMemberMapInfo(OmpMapParentAndMemberData &parentMembers,
+                              llvm::SmallVectorImpl<int64_t> &memberIndices) {
+  for (auto memberData : parentMembers.memberPlacementIndices)
+    if (std::equal(memberIndices.begin(), memberIndices.end(),
+                   memberData.begin()))
+      return true;
+  return false;
+}
+
+static void generateArrayIndices(lower::AbstractConverter &converter,
```
Comment (ergawy) [inline flang/lib/Lower/OpenMP/Utils.cpp:172]: Please add docs with an example.

File: flang/lib/Lower/OpenMP/Utils.cpp [hunk 8]
```diff
@@ -145,11 +146,174 @@ createMapInfoOp(fir::FirOpBuilder &builder, mlir::Location loc,
...
+      converter, firOpBuilder, *objectList[0].sym(), clauseLocation);
+  mlir::Value curValue = parentBaseAddr.addr;
+
+  // Iterate over all objects in the objectList, this should consist of all
+  // record types between the parent and the member being mapped (including
+  // the parent). The object list may also contain array objects as well,
+  // this can occur when specifying bounds or a specific element access
+  // within a member map, we skip these.
+  size_t currentIndex = 0;
+  for (size_t i = 0; i < objectList.size(); ++i) {
```
Comment (ergawy) [inline flang/lib/Lower/OpenMP/Utils.cpp:237]: We can use a range-based loop here.

File: flang/lib/Lower/OpenMP/Utils.cpp [hunk 9]
```diff
@@ -145,11 +146,174 @@ createMapInfoOp(fir::FirOpBuilder &builder, mlir::Location loc,
...
+// When mapping members of derived types, there is a chance that one of the
+// members along the way to a mapped member is an descriptor. In which case
+// we have to make sure we generate a map for those along the way otherwise
+// we will be missing a chunk of data required to actually map the member
+// type to device. This function effectively generates these maps and the
+// appropriate data accesses required to generate these maps. It will avoid
+// creating duplicate maps, as duplicates are just as bad as unmapped
+// descriptor data in a lot of cases for the runtime (and unnecessary
+// data movement should be avoided where possible)
+mlir::Value createParentSymAndGenIntermediateMaps(
```
Comment (ergawy) [inline flang/lib/Lower/OpenMP/Utils.cpp:210]: It is difficult for me to understand what this function does. Can you please document the paramemters (the ones specific to the function, no need to document `converter` and other usual parameters.

Can you also add some pseudo-MLIR guiding examples to the comments to help clarify the behavior of the function?

File: flang/lib/Lower/OpenMP/Utils.cpp [hunk 10]
```diff
@@ -183,99 +347,69 @@ getComponentObject(std::optional<Object> object,
...
 }
 
-static void
-generateMemberPlacementIndices(const Object &object,
-                               llvm::SmallVectorImpl<int> &indices,
-                               semantics::SemanticsContext &semaCtx) {
+void generateMemberPlacementIndices(const Object &object,
+                                    llvm::SmallVectorImpl<int64_t> &indices,
+                                    semantics::SemanticsContext &semaCtx) {
+  indices.clear();
```
Comment (ergawy) [inline flang/lib/Lower/OpenMP/Utils.cpp:353]: I think it would be better to assert the list is emtpy here.

File: flang/lib/Lower/OpenMP/Utils.cpp [hunk 11]
```diff
@@ -183,99 +347,69 @@ getComponentObject(std::optional<Object> object,
   return getComponentObject(baseObj.value(), semaCtx);
 }
 
-static void
-generateMemberPlacementIndices(const Object &object,
-                               llvm::SmallVectorImpl<int> &indices,
-                               semantics::SemanticsContext &semaCtx) {
+void generateMemberPlacementIndices(const Object &object,
+                                    llvm::SmallVectorImpl<int64_t> &indices,
```
Comment (ergawy) [inline flang/lib/Lower/OpenMP/Utils.cpp:351]: Just curious why the change from `int` to `int64_t` in particular?

**Review by ergawy — COMMENTED** [review]
Reviewed `Utils.h/.cpp`, submitting the current state of the review to prevent double-reviews when possible.

File: flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp [hunk 12]
```diff
@@ -225,32 +359,79 @@ class MapInfoFinalizationPass
...
+        if (mapVarsArr[i] == op) {
+          for (auto [j, mapMember] : llvm::enumerate(op.getMembers())) {
+            newMapOps.push_back(mapMember);
+            // for TargetOp's which have IsolatedFromAbove we must align the
+            // new additional map operand with an appropriate BlockArgument,
+            // as the printing and later processing currently requires a 1:1
+            // mapping of BlockArgs to MapInfoOp's at the same placement in
+            // each array (BlockArgs and MapVars).
+            if (directiveOp) {
+              directiveOp.getRegion().insertArgument(i + j, mapMember.getType(),
```
Comment (skatrak) [inline flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp:384]: This probably needs rewriting based on the `BlockArgOpenMPOpInterface`, since this index-based approach easily breaks if other entry block argument-generating clauses are introduced before `map` (such as `in_reduction`). This may be a helpful reference: https://github.com/llvm/llvm-project/blob/db1a76206902e6926464e30cc47249e217405dab/flang/lib/Lower/OpenMP/OpenMP.cpp#L1010-L1014

File: flang/lib/Lower/OpenMP/Utils.cpp [hunk 13]
```diff
@@ -13,19 +13,22 @@
 #include "Utils.h"
 
 #include "Clauses.h"
+#include <DirectivesCommon.h>
+
+#include <flang/Evaluate/fold.h>
 #include <flang/Lower/AbstractConverter.h>
+#include <flang/Lower/ConvertExprToHLFIR.h>
```
Comment (TIFitis) [inline flang/lib/Lower/OpenMP/Utils.cpp:20]: Are all of the header file changes necessary? Just making sure we only have the minimal set here.

File: flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp [hunk 14]
```diff
@@ -58,21 +67,62 @@ class MapInfoFinalizationPass
...
+  /// other MapInfoOp's and places them into the mapMemberUsers list, which
+  /// records the map that the current argument MapInfoOp "op" is part of
+  /// alongside the placement of "op" in the recorded users members list. The
+  /// intent of the generated list is to find all MapInfoOp's that may be
+  /// considered parents of the passed in "op" and in which it shows up in the
+  /// member list, alongside collecting the placement information of "op" in its
+  /// parents member list.
+  void
+  getMemberUserList(mlir::omp::MapInfoOp op,
+                    llvm::SmallVectorImpl<ParentAndPlacement> &mapMemberUsers) {
```
Comment (ergawy) [inline flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp:80]: In what scenario will `mapMemberUsers` end up with more than one element?

As a test (just to learn more about the changes), I modified this function as follows and none of the tests failed:
```diff
--- a/flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp
+++ b/flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp
@@ -81,8 +81,10 @@ class MapInfoFinalizationPass
     for (auto *users : op->getUsers())
       if (auto map = mlir::dyn_cast_if_present<mlir::omp::MapInfoOp>(users))
         for (auto [i, mapMember] : llvm::enumerate(map.getMembers()))
-          if (mapMember.getDefiningOp() == op)
+          if (mapMember.getDefiningOp() == op) {
             mapMemberUsers.push_back({map, i});
+            break;
+          }
   }
 
   llvm::SmallVector<int64_t>
```

Also, `mapMemberUsers[0]` is the only element used below (in `genDescriptorMemberMaps`).

**Review by TIFitis — APPROVED** [review]
LGTM :)

File: flang/lib/Lower/OpenMP/Utils.cpp [hunk 15]
```diff
@@ -145,11 +146,174 @@ createMapInfoOp(fir::FirOpBuilder &builder, mlir::Location loc,
...
+  return omp::ObjectList{llvm::reverse(objList)};
+}
+
+bool isDuplicateMemberMapInfo(OmpMapParentAndMemberData &parentMembers,
+                              llvm::SmallVectorImpl<int64_t> &memberIndices) {
+  for (auto memberData : parentMembers.memberPlacementIndices)
+    if (std::equal(memberIndices.begin(), memberIndices.end(),
+                   memberData.begin()))
+      return true;
+  return false;
```
Comment (skatrak) [inline flang/lib/Lower/OpenMP/Utils.cpp:169]: I think this can be simplified a bit further with something similar to this:
```suggestion
  return llvm::find_if(parentMembers.memberPlacementIndices,
      [&](auto &memberData) {
        return llvm::equal(memberIndices, memberData);
      }) != parentMembers.memberPlacementIndices.end();
```

**Review by skatrak — COMMENTED** [review]
Thank you Andrew for your work on this. I have a few relatively minor comments, but otherwise the general approach seems ok to me.

**Review by agozillon — COMMENTED** [review]
Unfortunately had to create a new iteration of PRs for this stack, as I moved machine and SPR decided that it didn't like that for some reason or other, but I've moved away from using SPR for the moment and did the new ones "by hand".

So I am incredibly sorry about that and the spam. I've submitted review comments here in reply, which unfortunately cannot be carried over (wish I could do so trivially), otherwise the new PR stack has incorporated all the changes I could.
